llvm/lib/Target/AMDGPU/SIInstrInfo.cpp

```diff
@@ -171,6 +171,39 @@ bool SIInstrInfo::isIgnorableUse(const MachineOperand &MO) const {
          isVALU(*MO.getParent()) && !resultDependsOnExec(*MO.getParent());
 }
 
+bool SIInstrInfo::modifiesRegisterImplicitly(
```

Comment:
We have an ever growing set of convoluted callbacks to deal with this problem. I think we need to look into moving some of the modeling of this into the CFG itself. I think we need to mark block edges as convergent/divergent/something